---
title: "Creating location plots"
output: html_notebook
---

The following code is for generating the plots showing locations of mutations across the genome in multiple samples, and for the analysis of mutation clustering across genotypes.

Libraries
```{r}
library(VariantAnnotation)
source('useful_functions.R')
source('plotting_functions.R')
library(ggplot2)
library(reshape2)
library(openxlsx)
```

Read sample annotation
```{r}
data <- read.xlsx("data/MA_paper_supplementary_table_1.xlsx", sheet = 2)
data$Sample <- as.character(data$Sample)
data$Genotype <- as.character(data$Genotype)
data$Code <- paste0(data$Genotype,':', data$Generation, sep = '')
# remove reference sample from further analysis
data <- data[-match('CD0001b',data$Sample),]
```

Read the VCF lists with base substitutions, indels and SVs
```{r}
# Upload indels
indels_dedup <- sapply(data$Sample, function(x) readVcf(paste0('PATH/TO/FILTERED/INDEL/VCFS/',x,'.vcf')))
# Upload subs
vcfs_dedup <- sapply(data$Sample, function(x) readVcf(paste0('PATH/TO/FILTERED/SNV/VCFS/',x,'.vcf')))
# Upload SVs
SVclust.new <- sapply(data$Sample, function(x) read.table(paste0('PATH/TO/FILTERED/SV/TABLES/',x,'.tsv')))
```

Select only intrachromosomal SVs and merge clusters of SVs
```{r}
svmat <- do.call('rbind',SVclust.new)
svmat <- svmat[as.character(svmat$CHR1) == as.character(svmat$CHR2),]
svmat$CHR1 <- as.character(svmat$CHR1);svmat$CHR2 <- as.character(svmat$CHR2);svmat$Sample <- as.character(svmat$Sample)
svmat.new <- svmat[,c(1:4,12:13)]
svmat <- svmat[svmat$clust.type != 'some',]
for (w in unique(svmat$Sample)) {
  tmp <- svmat[svmat$Sample == w,]
  tmp1 <- tmp[,c(1:4,12:13)]
  for (j in unique(tmp$clust)) {
    clust.tmp <- tmp[tmp$clust == j,]
    tmp1 <- rbind(tmp1,
                  c(as.character(clust.tmp$CHR1[1]), min(as.numeric(clust.tmp$POS1)), 
                    as.character(clust.tmp$CHR2[1]), max(as.numeric(clust.tmp$POS2)),
                    clust.tmp$Sample[1], clust.tmp$clust.type[1]))
  }
  tmp1 <- tmp1[-c(1:nrow(tmp)),,drop = F]
  svmat.new <- svmat.new[-which(svmat.new$Sample == w),]
  svmat.new <- rbind(svmat.new, tmp1)
  print(w)
}
svmat.new$POS1 <- as.numeric(svmat.new$POS1)
svmat.new$POS2 <- as.numeric(svmat.new$POS2)
svmat <- svmat.new
```

Prepare the genome data
```{r}
chr_lens <- c(15072434,15279421,13783801,17493829,20924180,17718942)
names(chr_lens) <- c('I','II','III','IV','V','X')
df <- data.frame(name = names(chr_lens), length = chr_lens)
```

Function to check clustering the genome. We assessed it using the start points of all the base substitutions and indels across the samples of the same genotype and generation. The function checks if a variant has another base change or indel within 100 bp from it and assigns a clustered status if so.

```{r}
isClustered <- function(variant_table, window = 1000){
  res <- rep(0, nrow(variant_table))
  d_next <- c(diff(variant_table$start),window+1)
  d_prev <- c(window+1, diff(variant_table$start))
  res[ (d_next < 100 & c(diff(as.numeric(variant_table$seqnames)) == 0, FALSE)) |
         (d_prev < 100 & c(FALSE, diff(as.numeric(variant_table$seqnames)) == 0)) ] <- 1
  return(res)
}
```


Checking if a variant is an MNV
```{r}
isMNV <- function(vcf) {
  d <- diff(start(vcf)) == 1 & abs(diff(geno(vcf)$PM[,"TUMOUR"] )) <= 0.05
  w <- c(FALSE, d) | c(d, FALSE)
  return(w)
}
```

*Location plots*

Start generating images
```{r}
experiments <- split(data$Sample, data$Code)
huge_tot <- list()  
totcount <- 1
q <- list()
j <- 1
clrs <- c("#2EBAED","#000000","#DE1C14","#D4D2D2","#ADCC54","#F0D0CE","brown","#8DD3C7","#FFFFB3","#BEBADA")
names(clrs) <- c('C>A','C>G','C>T','T>A','T>C','T>G','DNV','D','DI','I')
for (gntp in unique(data$Genotype)) {
  
  inds <- names(experiments)[grep(gntp, names(experiments))]
  
  for (exp in inds) {
  
    samset <- experiments[[exp]]
    
    tot <- list()
    
    no.samples <- length(samset)
    
    for (n in samset) {

      tmp_vcf <- vcfs_dedup[[n]]
      if (length(tmp_vcf) > 0) {
        #mcols(tmp_vcf)$Sample <- n
        tmp <- isMNV(tmp_vcf)
        tmp_vcf_df <- as.data.frame(granges(tmp_vcf))
        tmp_vcf_df$Sample <- n
        tmp_vcf_df$VAF <- geno(tmp_vcf)[['PM']][,2]
        tmp_vcf_df$Type <- NA
        if (sum(tmp)>0) {
          dnvs <- tmp_vcf_df[tmp,][seq(1,sum(tmp),2),]
          dnvs$Type <- 'DNV'
        } else {
          dnvs <- tmp_vcf_df[tmp,]
        }
        subs <- tmp_vcf_df[!tmp,]
        if (length(subs) >0) {
          subs <- subs[subs$seqnames!='MtDNA',]
          subs$Type <- paste0(subs$REF,'>',unlist(sapply(subs$ALT,as.character)))
          subs$Type[subs$Type == 'A>C'] <- 'T>G'
          subs$Type[subs$Type == 'A>G'] <- 'T>C'
          subs$Type[subs$Type == 'A>T'] <- 'T>A'
          subs$Type[subs$Type == 'G>A'] <- 'C>T'
          subs$Type[subs$Type == 'G>C'] <- 'C>G'
          subs$Type[subs$Type == 'G>T'] <- 'C>A'
        }
        if (nrow(subs) == 0) subs <- NULL
        if (nrow(dnvs) == 0) dnvs <- NULL
      } else {
        subs <- NULL #as.data.frame(granges(tmp_vcf))
        dnvs <- NULL #as.data.frame(granges(tmp_vcf))
      }
      
  
      vcf <- indels_dedup[[n]]
      if (length(vcf) >0) {
        indels <- as.data.frame(granges(vcf))
        if (nrow(indels) > 0) {
          indels <- indels[indels$seqnames!='MtDNA',]
          indels$Sample <- n
          indels$VAF <- (geno(vcf)$PU[,"TUMOUR"] + geno(vcf)$NU[,"TUMOUR"]) / 
            (geno(vcf)$PR[,"TUMOUR"] + geno(vcf)$NR[,"TUMOUR"]) 
          indels$Type <- ifelse(nchar(indels$REF) > nchar(unlist(sapply(indels$ALT,as.character))), yes = 'D', no = 'I')
          indels$Type[nchar(indels$REF) > 1 & nchar(unlist(sapply(indels$ALT,as.character))) > 1] <- 'DI'
        } else indels <- NULL
      } else indels <- NULL
  
      if (!is.null(subs)) {
        if (!is.null(indels)) {
          if (!is.null(dnvs)) tot[[n]] <- rbind(subs,indels,dnvs)
          else tot[[n]] <- rbind(subs,indels)
        } else {
          if (!is.null(dnvs)) tot[[n]] <- rbind(subs,dnvs)
          else tot[[n]] <- subs
        }
      } else {
        if (!is.null(indels)) {
          if (!is.null(dnvs)) tot[[n]] <- rbind(indels,dnvs)
          else tot[[n]] <- indels
        } else {
          if (!is.null(dnvs)) tot[[n]] <- rbind(dnvs)
          else tot[[n]] <- NULL
        }
      }

      if (length(tot[[n]])>0) {
        tot[[n]]$Mode <- NA
        tot[[n]]$experiment <- data$Code[data$Sample == n]
        if (!is.null(subs)) {
          tot[[n]]$Mode[1:nrow(subs)] <- 'A'
          nsubs <- nrow(subs)
        } else nsubs <- 0
        if (!is.null(indels)) {
          tot[[n]]$Mode[(nsubs+1):(nsubs + nrow(indels))] <- 'B'
          ninds <- nrow(indels)
        } else ninds <- 0
        if (!is.null(dnvs)) {
          tot[[n]]$Mode[(nsubs+ninds+1):nrow(tot[[n]])] <- 'C'
        }
        rownames(tot[[n]]) <- NULL
        tot[[n]] <- tot[[n]][order(tot[[n]]$seqnames),]
        for (ch in levels(tot[[n]]$seqnames)) {
          tot[[n]][tot[[n]]$seqnames == ch,] <- tot[[n]][tot[[n]]$seqnames == ch,][order(tot[[n]]$start[tot[[n]]$seqnames == ch]),]
        }
        if (max(table(tot[[n]]$seqnames)) < 3) {
          tot[[n]]$clust <- 2
        } else {
          k <- isClustered(tot[[n]])
          tot[[n]]$clust <- k*3 + 2
        }
      }
    }
    tot <- do.call('rbind', tot)
    huge_tot[[exp]] <- tot
    
    # remove duplicates for plotting
    for (lll in unique(tot$start)) {
      if (sum(tot$start == lll) == 1) next
      if (var(tot$VAF[tot$start == lll]) < 0.01 & length(unique(tot$Type[tot$start == lll])) == 1)
        tot <- tot[-which(tot$start == lll)[-1],]
    }  

    # plotting with clusters
    if (sum(tot$clust>2)>0)
      q[[j]] <- ggplot() + geom_bar(data = df, aes(x = name, y = length), stat = 'identity',fill = 'white') +
        scale_y_continuous(labels = c('0 Mb','5 Mb', '10 Mb','15 Mb','20 Mb')) +
        geom_jitter(data = tot, aes(x = seqnames, y = start, col = Type, shape =Mode, size = clust)) +
        labs(title = paste0('Mutations across all samples from ',data$Code[data$Sample == samset[1]], ' experiment, ',
                          no.samples, ' samples'),
             x = 'Chromosome',y='Position') +
        scale_shape_discrete(labels = c('Substitutions','Indels','DNVs')) +
        scale_size_manual(values = c(2,5), labels = c('single','clustered')) +
        scale_color_manual(values=clrs[unique(tot$Type)]) +
      guides(size=guide_legend(title="Clustering"), shape = guide_legend(title='Class')) +
        theme(title = element_text(size=10))
    else # plotting without clusters
      q[[j]] <- ggplot() + geom_bar(data = df, aes(x = name, y = length), stat = 'identity',fill = 'white') +
      scale_y_continuous(labels = c('0 Mb','5 Mb', '10 Mb','15 Mb','20 Mb')) +
      geom_jitter(data = tot, aes(x = seqnames, y = start, col = Type, shape = Mode), size = 2) +
      labs(title = paste0('Mutations across all samples from ',data$Code[data$Sample == samset[1]], ' experiment, ',
                          no.samples, ' samples'),
           x = 'Chromosome',y='Position') +
      scale_shape_discrete(labels = c('Substitutions','Indels','DNVs')) +
      scale_color_manual(values=clrs[unique(tot$Type)]) +
      guides(shape = guide_legend(title='Class')) +
      theme(title = element_text(size=10))

    # plotting SVs
    COLOR=c("#FBB4AE","#B3CDE3","#CCEBC5","#DECBE4","#FED9A6","#FFFFCC","#E5D8BD")
    names(COLOR) = c("TD", "DEL", "INV", "COMPLEX", "FOLDBACK", "MOVE", "TRSL")
    svs <- svmat[as.character(svmat$Sample) %in% samset,]
    svs$POS1 <- as.numeric(svs$POS1)
    svs$POS2 <- as.numeric(svs$POS2)
    svs$CHR1 <- match(svs$CHR1, df$name)
    svs$clust.type <- factor(svs$clust.type)

    if (nrow(svs)>0) {
        q[[j]] <- q[[j]] + geom_rect(data = svs,
                                     mapping = aes(xmin = CHR1-0.5,
                                                   xmax = CHR1+0.5,
                                                   ymin = POS1,
                                                   ymax = POS2+10000,
                                                   fill = clust.type),
                                     alpha = 0.4) +
        scale_fill_manual(values = COLOR) +
          guides(fill = guide_legend(title='SV'))
    }

    j = j+1
    print(j)
  }
}
```

Save them all
```{r}
GG_save_pdf(q, filename = 'Mutation_accumulation_location_plots.pdf', 8,7)
```

Read in mutation counts and generate barplots of mutation types per sample (no counts for reference sample, CD0001b):
```{r}
Y <- read.csv('data/MA_mutation_counts.csv', row.names = 1)
short.Y <- cbind(rowSums(Y[,1:16]),rowSums(Y[,17:32]),rowSums(Y[,33:48]),rowSums(Y[,49:64]),
                 rowSums(Y[,65:80]),rowSums(Y[,81:96]), rowSums(Y[,97:98]), rowSums(Y[,99:104]),
                 rowSums(Y[,105:106]), rowSums(Y[,107:112]),rowSums(Y[,113:119]))
```

BER and NER baplot
```{r}
pdf('Figure_3b.pdf',10,6)
f <- barplot(t(rbind(short.Y[data$Sample[data$Code== 'N2:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'N2:20'],],
                     short.Y[data$Sample[data$Code== 'N2:40'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'agt-2:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'agt-2:20'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'ung-1:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'ung-1:20'],],
                     short.Y[data$Sample[data$Code== 'ung-1:40'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'xpa-1:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'xpa-1:20'],],
                     short.Y[data$Sample[data$Code== 'xpa-1:40'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'xpf-1:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'xpf-1:20'],],
                     short.Y[data$Sample[data$Code== 'xpf-1:40'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'xpg-1:1'],,drop = F],
                     short.Y[data$Sample[data$Code== 'xpg-1:20'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'xpc-1:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'xpc-1:20'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'csb-1:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'csb-1:20'],])),
             col = c("#2EBAED","#000000","#DE1C14","#D4D2D2","#ADCC54","#F0D0CE",
                     "brown","#8DD3C7","#FFFFB3","#BEBADA","darkmagenta"),
             xaxt = 'n', las = 2, border = NA, ylim = c(-2,50))
lines(x = rep(f[4] + (f[5]-f[4])/2,2), y = c(-2,30))
lines(x = rep(f[23] + (f[24]-f[23])/2,2), y = c(-2,40))
lines(x = rep(f[31] + (f[32]-f[31])/2,2), y = c(-2,40))
lines(x = rep(f[44] + (f[45]-f[44])/2,2), y = c(-2,40))
lines(x = rep(f[47] + (f[48]-f[47])/2,2), y = c(-2,40))
lines(x = rep(f[54] + (f[55]-f[54])/2,2), y = c(-2,40))
lines(x = rep(f[57] + (f[48]-f[47])/2,2), y = c(-2,40))
lines(x = rep(f[64] + (f[65]-f[64])/2,2), y = c(-2,40))
lines(x = rep(f[67] + (f[68]-f[67])/2,2), y = c(-2,40))
lines(x = rep(f[75] + (f[76]-f[75])/2,2), y = c(-2,40))
lines(x = rep(f[84] + (f[85]-f[84])/2,2), y = c(-2,40))
lines(x = rep(f[96] + (f[97]-f[96])/2,2), y = c(-2,40))
legend('topleft', fill = c("#2EBAED","#000000","#DE1C14","#D4D2D2","#ADCC54","#F0D0CE",
                           "brown","#8DD3C7","#FFFFB3","#BEBADA","darkmagenta"),ncol = 2,
       legend = c('C>A','C>G','C>T','T>A','T>C','T>G','MNV', 
                  'Del', 'Del/Ins', 'Ins', 'SV'), border = NA, bty = 'n')
dev.off()
```

BRCA
```{r}
pdf('brca_barplots.pdf',10,6)
f <- barplot(t(rbind(short.Y[data$Sample[data$Code== 'ced-3:1'],,drop = F],
                     short.Y[data$Sample[data$Code== 'ced-3:20'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'cep-1:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'cep-1:1'],,drop = F],
                     short.Y[data$Sample[data$Code== 'cep-1:20'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*:1'],,drop = F],
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*:20'],],
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*:40'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*;ced-3:0'],,drop = F],
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*;ced-3:1'],,drop = F],
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*;ced-3:20'],],
                     NA,
                     NA,
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*; cep-1:10'],,drop = F],
                     short.Y[data$Sample[data$Code== 'brc-1 brd-1*; cep-1:20'],])), 
             col = c("#2EBAED","#000000","#DE1C14","#D4D2D2","#ADCC54","#F0D0CE",
                     "brown","#8DD3C7","#FFFFB3","#BEBADA","darkmagenta"),
             xaxt = 'n', las = 2, border = NA, ylim = c(-1,60))
lines(x = rep(f[3] + (f[4]-f[3])/2,2), y = c(-1,10), lwd =2)
lines(x = rep(f[14] + (f[15]-f[14])/2,2), y = c(-1,10), lwd =2)
lines(x = rep(f[16] + (f[17]-f[16])/2,2), y = c(-1,10), lwd =2)
lines(x = rep(f[25] + (f[26]-f[25])/2,2), y = c(-1,10), lwd =2)
lines(x = rep(f[27] + (f[28]-f[27])/2,2), y = c(-1,20), lwd =2)
lines(x = rep(f[34] + (f[35]-f[34])/2,2), y = c(-1,40), lwd =2)
lines(x = rep(f[45] + (f[46]-f[45])/2,2), y = c(-1,20), lwd =2)
legend('topleft', fill = c("#2EBAED","#000000","#DE1C14","#D4D2D2","#ADCC54","#F0D0CE",
                           "brown","#8DD3C7","#FFFFB3","#BEBADA","darkmagenta"),ncol = 2,
       legend = c('C>A','C>G','C>T','T>A','T>C','T>G','MNV', 
                  'Del', 'Del/Ins', 'Ins', 'SV'), border = NA, bty = 'n')
dev.off()
```

*Clustering analysis*

Aggregate all clusters
```{r}
allmuts <- do.call('rbind',huge_tot)
allclust <- allmuts[allmuts$clust>2,]

clustdf <- data.frame(name = unique(allclust$Sample))
clustdf$name <- as.character(clustdf$name)
clustdf$generation <- data$Generation[match(clustdf$name,data$Sample)]
clustdf$genotype <- data$Genotype[match(clustdf$name,data$Sample)]
clustdf$code <- data$Code[match(clustdf$name,data$Sample)]

clustdf$number_of_clusters <- sapply(clustdf$name, function(x) 
    sum(abs(diff(allclust$start[allclust$Sample == x]))>100) + 1)
clustdf$number_clust_muts <- sapply(clustdf$name, function(x) sum(allclust$Sample == x))
clustdf$total_samples_of_this_code <- sapply(clustdf$code, function(x) sum(data$Code == x))
clustdf$total_mutations_in_sample <- sapply(clustdf$name, function(x) sum(allmuts$Sample == x))
clustdf$prop.of.clust <- clustdf$number_clust_muts / clustdf$total_mutations_in_sample
```

Calculate clustering rates in genotypes with more than 3 samples of generation higher than 1, and compare to wild-type:

```{r}

df <- data.frame(sapply(sort(unique(data$Genotype)), function(x) 
  as.numeric(data$Genotype == x)))
rownames(df) <- data$Sample
df$prop_clust <- sapply(data$Sample, function(x) {
    if (x %in% clustdf$name) return(clustdf$prop.of.clust[clustdf$name == x])
    return(0)
})
df <- df[data$Generation>1,]
df <- df[,colSums(df>0)>2]

model <- glm(prop_clust ~ 0 + ., data = df)

summary(model)
coeffs <- coef(summary(model))

pvclust <- NULL
for (zzz in rownames(coeffs)[-match('N2',rownames(coeffs))]) {
  stat_mu = coeffs[zzz,1] - coeffs['N2',1]
  stat_sd = sqrt(coeffs[zzz,2]**2 + coeffs['N2',2]**2)
  zscore = stat_mu / stat_sd
  pvclust <- c(pvclust, 1 - pnorm(q = zscore))
}
rownames(coeffs)[-match('N2',rownames(coeffs))][which(p.adjust(pvclust,method='BH') < 0.05)]
names(pvclust) <- rownames(coeffs)[-match('N2',rownames(coeffs))]
pvprop <- p.adjust(pvclust,method='BH')
```

# Visualize
```{r}
pvprop <- c(pvprop[c(1:31)],1,pvprop[c(32:59)]) # insert N2 for simplicity
pdf('~/Cluster_comparison.pdf',6,6)
set.seed(123)
boxplot(coeffs[,1], frame = F, outline = F, ylim = c(0,max(coeffs[,1]+1.96*coeffs[,2])),
        ylab = 'Prop. of clustered mut-s', main = 'Proportion of clustered mutations\n across genotypes')
#newx2 <- jitter(rep(1, length(coeffs[,1])), amount = 0.1)
points(x = newx2, y = coeffs[,1], col = 'gray', pch = 16) 
o2 <- which(pvprop < 0.05)
points(x = newx2[o2], y = coeffs[o2,1], col = 'darkred', pch = 16) 
arrows(x0 = newx2[o2],y0 = coeffs[o2,1] - 1.96*coeffs[o2,2],
       y1=coeffs[o2,1]+1.96*coeffs[o2,2],
       col = 'gray21',lwd=0.5,length=0)
abline(h = coeffs['N2',1], lty = 2)
text(x = c(newx2[o2][1] - 0.1, newx2[o2][2] + 0.15, newx2[o2][3] + 0.13,newx2[o2][4] - 0.07,newx2[o2][5] - 0.05,newx2[o2][6] - 0.1),
     y = c(coeffs[o2,1][1:4], coeffs[o2,1][5]-0.01,coeffs[o2,1][6]+0.01), font = 3,
     labels = sapply(rownames(coeffs)[o2], function(x) data$Genotype[data$Sample == rownames(df)[df[,x]>0][1]]),
     cex = 0.7)
legend('topright', legend = 'significantly different\n from N2 (FDR 5%)', 
       col = 'darkred', bty = 'n', pch = 16, 
       border = NA)
dev.off()
```

Session information

```{r}
sessionInfo()
# R version 3.6.3 (2020-02-29)
# attached base packages:
#  [1] grid      stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] reshape2_1.4.3              ggplot2_3.2.1               VariantAnnotation_1.30.1    Rsamtools_2.0.3            
#  [5] Biostrings_2.52.0           XVector_0.24.0              SummarizedExperiment_1.14.1 DelayedArray_0.10.0        
#  [9] BiocParallel_1.18.1         matrixStats_0.55.0          Biobase_2.44.0              GenomicRanges_1.36.1       
# [13] GenomeInfoDb_1.20.0         IRanges_2.18.3              S4Vectors_0.22.1            BiocGenerics_0.30.0        
# 
# loaded via a namespace (and not attached):
#  [1] Rcpp_1.0.2               lattice_0.20-38          prettyunits_1.0.2        assertthat_0.2.1         digest_0.6.22          
#  [6] R6_2.4.0                 plyr_1.8.4               RSQLite_2.1.2            httr_1.4.1               pillar_1.4.2            
# [11] zlibbioc_1.30.0          rlang_0.4.5              GenomicFeatures_1.36.4   progress_1.2.2           lazyeval_0.2.2          
# [16] rstudioapi_0.10          blob_1.2.0               Matrix_1.2-18            stringr_1.4.0            RCurl_1.95-4.12         
# [21] bit_1.1-14               biomaRt_2.40.5           munsell_0.5.0            compiler_3.6.3           rtracklayer_1.44.4   
# [26] xfun_0.10                pkgconfig_2.0.3          tidyselect_0.2.5         tibble_2.1.3             GenomeInfoDbData_1.2.1  
# [31] XML_3.98-1.20            crayon_1.3.4             dplyr_0.8.3              withr_2.1.2              GenomicAlignments_1.20.1
# [36] bitops_1.0-6             gtable_0.3.0             DBI_1.0.0                magrittr_1.5             scales_1.0.0            
# [41] stringi_1.4.3            vctrs_0.2.4              tools_3.6.3              bit64_0.9-7              BSgenome_1.52.0         
# [46] glue_1.3.1               purrr_0.3.3              hms_0.5.3                AnnotationDbi_1.46.1     colorspace_1.4-1        
# [51] memoise_1.1.0            knitr_1.25 
```

